host/peerscore.go

```diff
@@ -0,0 +1,76 @@
+package host
+
+import (
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type TopicScore struct {
+	Topic                    string        `json:"topic"`
+	TimeInMesh               time.Duration `json:"timeInMesh"`
+	FirstMessageDeliveries   float64       `json:"firstMessageDeliveries"`
+	MeshMessageDeliveries    float64       `json:"meshMessageDeliveries"`
+	InvalidMessageDeliveries float64       `json:"invalidMessageDeliveries"`
+}
+
+type TraceEventPeerScore struct {
+	PeerID             string       `json:"peerID"`
+	Score              float64      `json:"score"`
+	AppSpecificScore   float64      `json:"appSpecificScore"`
+	IPColocationFactor float64      `json:"ipColocationFactor"`
+	BehaviourPenalty   float64      `json:"behaviourPenalty"`
+	Topics             []TopicScore `json:"topics"`
+}
+
+func composePeerScoreEventFromRawMap(pid peer.ID, score *pubsub.PeerScoreSnapshot) TraceEventPeerScore {
+	var topics []TopicScore
+	for topic, snapshot := range score.Topics {
+		topics = append(topics, TopicScore{
+			Topic:                    topic,
+			TimeInMesh:               snapshot.TimeInMesh,
+			FirstMessageDeliveries:   snapshot.FirstMessageDeliveries,
+			MeshMessageDeliveries:    snapshot.MeshMessageDeliveries,
+			InvalidMessageDeliveries: snapshot.InvalidMessageDeliveries,
+		})
+	}
+	return TraceEventPeerScore{
+		PeerID:             pid.String(),
+		Score:              score.Score,
+		AppSpecificScore:   score.AppSpecificScore,
+		IPColocationFactor: score.IPColocationFactor,
+		BehaviourPenalty:   score.BehaviourPenalty,
+		Topics:             topics,
+	}
+}
+
+const PeerScoreEventType = "PEERSCORE"
+
+// ScoreKeeper is a thread-safe local copy of the score per peer and per copy
+// TODO: figure out if this is some sort of info that we want to expose through OpenTelemetry (Still good to have it)
+type ScoreKeeper struct {
```

Comment:
The `Update()` method should directly swap the entire `map` for the one given by the GossipSub Inspect call.

from the [spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#peer-scoring)
> Furthermore, the score is retained for some period of time when a peer disconnects, so that malicious peers cannot easily reset their score when it drops to negative and well behaving peers don't lose their status because of a disconnection.

As far as I know, these scores are eventually dropped, so we would just be tracking all the scores available at the GossipSub level. (There seems to be a [RetainScore](https://github.com/libp2p/go-libp2p-pubsub/blob/048a4d30d0c3c00829181fd81aa697eb60497448/score_params.go#L111C2-L111C13) duration after the Disconnection)